swap/peer.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+type SwapPeer struct {
+	*protocols.Peer
+	swap *Swap
+}
+
+func NewPeer(p *protocols.Peer, s *Swap) *SwapPeer {
+	return &SwapPeer{
+		Peer: p,
+		swap: s,
+	}
+}
+
+func (sp *SwapPeer) handleMsg(ctx context.Context, msg interface{}) error {
+	switch msg := msg.(type) {
+
+	case *ChequeRequestMsg:
+		return sp.handleChequeRequestMsg(ctx, msg)
```

Comment:
Not directly related to this PR, but we have to consider if we want blocking or non-blocking processing of messages over devp2p protocols in Swarm.

Here you are handling them in a blocking way, which means that any other protocol will also block until this read/write unblocks.

Blocking is good because of back pressure (TCP flow control), but bad, because it can lead to deadlocks, and not efficient processing of messages - this is why in the `stream` protocol we mostly just read an incoming message and handle is asynchronously, immediately listening for new messages.

In my experience with Swarm, block IO has created numerous deadlocks and issues (I've personally debugged 2-3 such cases), so def worth considering early rather than later.

Reply:
Thanks for this heads-up. We will have to evaluate in this specific context (incentives) what makes more sense, async or sync.

Reply:
I agree that this has been a major headache. 
I suggest we go back to the original p2p/protocols design and let it register handles similar to pss. Then

* the handlers are by contract called non-blocking 
* we always call Send syncronously to handle errors properly
* there is only one way to drop a peer: by calling `Drop`

Reply:
@zelig what you suggest is one option:
- non-blocking , asynchronous handlers => we have no backpressure mechanism and we'd have to implement it separately at some point
- call Send synchronously and handle errors => yes, we must do that
- only one way to drop a peer, by calling `Drop` => ok

However you also mention the `p2p/protocols` design, which has another set of issues (or we use it incorrectly), so we should discuss this elsewhere as it is not directly relevant here.